src/tactic/chain_trans.lean

```diff
@@ -0,0 +1,170 @@
+/-
+Copyright (c) 2020 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Simon Hudon
+-/
+import meta.rb_map
+import meta.expr
+import tactic.core
+/-!
+# `chain_trans` tactic
+
+This file defines the `chain_trans` tactic which prove goals of the
+shape `x ≤ y` or `x < y` by finding assumptions of the shape `x ≤ z`,
+`x < z` or `x = z` and combining them using transitivity. In order
+to prove `x < y`, at least one of the assumptions must be
+a strict ordering. Whatever is found, `x ≤ y` can be proved.
+
+```lean
+variables {w x y z : ℕ}
+example (h₀ : w ≤ x) (h₁ : x = y) (h₂ : y < z) : w < z :=
+begin
+  chain_trans,
+end
+```
+
+## Implementation
+
+`chain_trans` builds a graph where vertices are expressions of type
+`α`, with a preorder on `α`, and edges are assumptions stating an
+order between its source and its target. It then uses depth first
+search to find a path through that graph between two vertices.
+the edges along that path can then be combined through transitivity
+to prove an order between the two vertices.
+-/
+
+universes u v
+
+namespace tactic
+
+open expr
+
+namespace chain_trans
+
+/-- A label for the edges of a graph of assumptions to
+specify whether the assumption is a strict ordering
+(i.e. `<`) or non-strict ordering (i.e. `≤`). -/
+@[derive [inhabited, decidable_eq]]
+inductive edge
+| lt | le
+
+instance : has_to_string edge :=
+⟨ λ e, match e with
+      | edge.lt := "lt"
+      | edge.le := "le"
+      end ⟩
+
+meta instance edge.has_to_format : has_to_format edge :=
+⟨ λ e, to_fmt $ to_string e ⟩
+
+open native
+
+/--
+Place `lt` edges first in edge list.
+-/
+meta def sort_edges (ls : list (expr × edge × expr)) : list (expr × edge × expr) :=
+let ⟨xs₀, xs₁⟩ := ls.partition (λ e : _ × _ × _, e.2.1 = edge.lt) in
+xs₀ ++ xs₁
+
+/-- A graph, represented as adjacency lists, where edges are
+expressions of type `α` and where an edge between `u` and `v` contains
+a proof that `u ≤ v` or `u < v`.
+
+Each edge is stored as a triple (proof, label, target) in
+the list of edges of its source. The label is `lt` or `le` and
+the proof establishes the appropriate relation between the source
+and the target of the edge.
+
+The list of edges of each vertex is sorted so as to keep the
+edges labelled `lt` first in the list. This ensures that,
+when traversing the graph, paths that prove a stronger relation
+are going to be considered first. -/
+meta def graph := native.rb_lmap expr (expr × edge × expr)
+
+/--
+Construct a `graph` from a list of edges. Edges are represented as
+tuples `(proof, label, source, target)`.
+-/
+meta def graph.mk (es : list (expr × edge × expr × expr)) : graph :=
+rb_map.map sort_edges $ es.foldl (λ (g : graph) ⟨pr,e,x,y⟩, g.insert x (pr,e,y)) (rb_lmap.mk _ _)
+
+meta instance graph.has_to_format : has_to_tactic_format graph :=
+by delta graph; apply_instance
+
+private meta def dfs_trans' (g : graph) (r : ref expr_set) (v : expr) :
```

Comment:
As I read it, this is building proofs during the search, right? So in a big problem, it will build lots of transitivity terms that end up getting discarded? It might be better to store lists of transitive edges during the search and build the proof term at the end. Based on my experiences, I bet the costs of elaboration in using `mk_app` will be most of the cost of the search. (But I haven't profiled anything here yet.)

Reply:
It started off like you describe (building the proof as I go) but I switched to only accumulating the edges and chaining them up when I know for sure that this is the proof that I want.

Reply:
Could you write docs for this function then? It's hard for me to recognize that without knowing what the inputs are.